File: Assembly-FSharp/Assembly-FSharp/Managers/RagdollManager.fs
namespace Styx.Managers

open UnityEngine

type RagdollState =
        | Ragdolled
        | Animated

type RagdollManager () =
    inherit MonoBehaviour ()

    let mutable state: RagdollState = Animated
    [<SerializeField>]
    let mutable rigidbodies: Rigidbody[] = [||]

    [<SerializeField>]
    let mutable animator: Animator = Unchecked.defaultof<Animator>
    [<SerializeField>]
    let mutable rbody: Rigidbody = Unchecked.defaultof<Rigidbody>
    [<SerializeField>]
    let mutable mainCollider: CapsuleCollider = Unchecked.defaultof<CapsuleCollider>

    let setKinematic (value:bool) (rigidbody: Rigidbody) =
        rigidbody.isKinematic <- value


    [<SerializeField>]
    member this.Ragdolled
        with get() = (state = Ragdolled)
        and set(v) = 
            match v with
            | true ->
                Array.iter (setKinematic false) rigidbodies
                rbody.isKinematic <- true
                animator.enabled <- false
                mainCollider.enabled <- false
                state <- Ragdolled
            | false ->
                Array.iter (setKinematic true) rigidbodies
                rbody.isKinematic <- false
                animator.enabled <- true
                mainCollider.enabled <- true
                state <- Animated
                

    member this.Start() =
        rigidbodies <- this.GetComponentsInChildren<Rigidbody>()
        this.Ragdolled <- false


    member this.FixedUpdate() =
        this.Ragdolled <- not animator.isActiveAndEnabled

File: Assets/_Scripts/Managers/RagdollManager.cs
using Styx.Models;
using System.Linq;
using UnityEngine;

namespace Assets._Scripts.Managers
{
    public class RagdollManager : MonoBehaviour
    {
        [SerializeField]
        private RagdollModel _ragdollModel;

        public void Start()
        {
            _ragdollModel.MainAnimator = (_ragdollModel.MainAnimator ?? GetComponent<Animator>());
            _ragdollModel.MainRigidbody = (_ragdollModel.MainRigidbody ?? GetComponent<Rigidbody>());
            _ragdollModel.MainCollider = (_ragdollModel.MainCollider ?? GetComponent<CapsuleCollider>());

            _ragdollModel.Rigidbodies = GetComponentsInChildren<Rigidbody>().ToList();
            _ragdollModel.Colliders = GetComponentsInChildren<Collider>().ToList();
            Ragdolled = false;
        }

        public bool Ragdolled {
            get { return _ragdollModel.State == RagdollState.Ragdolled; }
            set
            {
                if (value == true)
                {
                    _ragdollModel.Rigidbodies.ForEach(r => r.isKinematic = false);
                    _ragdollModel.Colliders.ForEach(c => c.enabled = true);
                    _ragdollModel.MainRigidbody.isKinematic = true;
                    _ragdollModel.MainAnimator.enabled = false;
                    _ragdollModel.MainCollider.enabled = false;
                    _ragdollModel.State = RagdollState.Ragdolled;
                }
                else
                {
                    _ragdollModel.Rigidbodies.ForEach(r => r.isKinematic = true);
                    _ragdollModel.Colliders.ForEach(c => c.enabled = false);
                    _ragdollModel.MainRigidbody.isKinematic = false;
                    _ragdollModel.MainAnimator.enabled = true;
                    _ragdollModel.MainCollider.enabled = true;
                    _ragdollModel.State = RagdollState.Animated;
                }
            }
        }

        public void FixedUpdate()
        {
            if (Ragdolled == _ragdollModel.MainAnimator.isActiveAndEnabled)
                Ragdolled = !_ragdollModel.MainAnimator.isActiveAndEnabled;
        }
    }
}
